package/libskylark/src/libskylark.c

```diff
@@ -0,0 +1,341 @@
+      NO_ERROR) {
+    return rc;
+  }
+  config->num_retries = 0;
+  config->retry_delay = 0;
+  config->retry_max_time = 0;
+  config->fd = 0;
+  memset(config->endpoint_url, '\0', BUFSIZE);
+  return rc;
+}
+
+/** Setup libcurl global variables.
+ *
+ * \return  RC return code indicating success or failure
+ */
+RC setup_globals(void)
+{
+  CURLcode res = curl_global_init(CURL_GLOBAL_ALL);
+  if (res != CURLE_OK) {
+    log_error("%d %s\n", res, curl_easy_strerror(res));
+    return -E_INITIALIZATION_ERROR;
+  }
+  return NO_ERROR;
+}
+
+/** Cleanup libcurl global variables.
+ *
+ * \return  RC return code indicating success or failure
+ */
+void teardown_globals(void) { curl_global_cleanup(); }
+/**
+ *  Download (i.e., rover) processes.
+ */
+
+/** Callback for the download process: writes to a file descriptor.
+ *
+ * For more details, see:
+ * https://curl.haxx.se/libcurl/c/CURLOPT_WRITEFUNCTION.html
+ *
+ * \param ptr       Pointer to the data received by libcurl
+ * \param size      Item size
+ * \param nmemb     Number of items
+ * \param userdata  Pointer to destination
+ * \return RC return code indicating success or failure
+ */
+size_t download_callback(char *ptr, size_t size, size_t nmemb, void *userdata)
+{
+  int *fd = (int *)userdata;
+  ssize_t m = write(*fd, ptr, size * nmemb);
+#ifdef VERBOSE
+  log_error("download_callback! size=%d items=%d\n", (int)size, (int)nmemb);
+#endif
+  return m;
+}
+
+/** libcurl-based subscribe request; streams SBP over a continuous connection
+ * and writes that data to a file descriptor passed in the client
+ * configuration. The expectation here is that a reader of that file descriptor
+ * is passing it on to the firmware. The write to the file descriptor is handled
+ * via the passed callback function.
+ *
+ * Dataflow:
+ *   HTTP GET => callback writer => config->fd => reader in external process
+ *
+ * \param config   Skylark Client configuration
+ * \param cb       Callback function writing to a file descriptor.
+ * \param verbose  Verbose output
+ * \return  RC return code indicating success or failure
+ */
+RC download_process(client_config_t *config, write_callback_fn cb, bool verbose)
+{
+  log_error("Start of download!\n");
+  CURL *curl = curl_easy_init();
+  if (curl == NULL) {
+    return -E_NULL_VALUE_ERROR;
+  }
+  curl_easy_setopt(curl, CURLOPT_URL, config->endpoint_url);
+  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, cb);
+  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &config->fd);
+  if (verbose) {
+    curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+  }
+  curl_easy_setopt(curl, CURLOPT_USERAGENT, USER_AGENT);
+  struct curl_slist *chunk = NULL;
+  chunk = curl_slist_append(chunk, STREAM_ENCODING);
+  chunk = curl_slist_append(chunk, SBP_V2_ACCEPT_TYPE);
+  chunk = curl_slist_append(chunk, config->device_header);
+  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, chunk);
+  CURLcode res = curl_easy_perform(curl);
+  if (res != CURLE_OK) {
```

Comment:
This process needs to be updated to be like the example below. Note that the handling of HTTP status codes will be different.